### ISSUE TITLE
Missing checks for existential types leads to unsoundness

### ISSUE THREAD

**Issue by smarter** [lampepfl/dotty#583]
this should not compile:
```scala
import scala.collection.mutable.ListBuffer

object Test {
  def main(args: Array[String]): Unit = {
    val f: ListBuffer[Int] = ListBuffer(1,2)
    val g: ListBuffer[Double] = ListBuffer(3.0,4.0)
    val lb: ListBuffer[ListBuffer[_ <: AnyVal]] = ListBuffer(f, g)
    lb(0)(0) = lb(1)(0)
    val x: Int = f(0) // ClassCastException: java.lang.Double cannot be cast to java.lang.Integer
  }
}
```
In `scalac` it fails with:
```scala
t583.scala:8: error: type mismatch;
 found   : (some other)_$1(in method main) where type (some other)_$1(in method main) <: AnyVal
 required: _$1(in method main) where type _$1(in method main) <: AnyVal
    lb(0)(0) = lb(1)(0)
                    ^
one error found
```

**Comment by odersky** [comment]
Yes, there's some logic that still needs to be implemented here.

**Comment by smarter** [comment]
Fixed in #599.
